src/authentication/authentication-hapi-plugin.ts

```diff
@@ -0,0 +1,154 @@
+
+import * as Boom from 'boom';
+import * as auth from 'hapi-auth-jwt2';
+import { inject, injectable } from 'inversify';
+import * as Joi from 'joi';
+import * as qs from 'querystring';
+
+import * as Hapi from '../server/hapi';
+import { HapiPlugin } from '../server/hapi-register';
+import { Group, User } from '../shared/gitlab';
+import { GitlabClient } from '../shared/gitlab-client';
+import { Logger, loggerInjectSymbol } from '../shared/logger';
+import {default as AuthenticationModule} from './authentication-module';
+import { AccessToken, authServerBaseUrlInjectSymbol, jwtOptionsInjectSymbol } from './types';
+
+interface UserId {
+  id: string;
+  idp: string;
+}
+
+@injectable()
+class AuthenticationHapiPlugin extends HapiPlugin {
+
+  public static injectSymbol = Symbol('authentication-hapi-plugin');
+  private readonly authenticationModule: AuthenticationModule;
+  private readonly authServerBaseUrl: string;
+  private readonly logger: Logger;
+  private readonly gitlab: GitlabClient;
+  private readonly options: auth.JWTStrategyOptions;
+  private teamTokenClaimKey: 'https://team_token' = 'https://team_token';
+  private subEmailClaimKey: 'https://sub_email' = 'https://sub_email';
+
+  constructor(
+    @inject(AuthenticationModule.injectSymbol) authenticationModule: AuthenticationModule,
+    @inject(authServerBaseUrlInjectSymbol) authServerBaseUrl: string,
+    @inject(loggerInjectSymbol) logger: Logger,
+    @inject(GitlabClient.injectSymbol) gitlab: GitlabClient,
+    @inject(jwtOptionsInjectSymbol) jwtOptions: auth.JWTStrategyOptions,
+    ) {
+    super({
+      name: 'authentication-plugin',
+      version: '1.0.0',
+    });
+    this.authenticationModule = authenticationModule;
+    this.authServerBaseUrl = authServerBaseUrl;
+    this.logger = logger;
+    this.gitlab = gitlab;
+    this.options = jwtOptions;
+  }
+
+  public async register(server: Hapi.Server, _options: Hapi.IServerOptions, next: () => void) {
+    await this.registerAuth(server);
+    server.route([{
+      method: 'GET',
+      path: '/signup/{teamToken}',
```

Comment:
Yes, I was for some reason thinking about the initial request that the user makes. At least when writing the route.